File: server/service/tts/sentence_splitter_parity_test.go
package tts

import (
	"encoding/json"
	"os"
	"testing"
)

// TestSplitterParity asserts the Go splitter matches the shared fixture file
// at testdata/splitter_parity.json. The Dart splitter test reads the same
// file. If either side changes the splitting algorithm, both tests must be
// updated in lockstep — that's the entire point of this fixture.
func TestSplitterParity(t *testing.T) {
	type expected struct {
		Text       string `json:"text"`
		CharOffset int    `json:"charOffset"`
	}
	type fixCase struct {
		Name     string     `json:"name"`
		Input    string     `json:"input"`
		Splitter *Splitter  `json:"splitter,omitempty"`
		Expected []expected `json:"expected"`
	}
	type fixture struct {
		Cases []fixCase `json:"cases"`
	}

	raw, err := os.ReadFile("testdata/splitter_parity.json")
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fix fixture
	if err := json.Unmarshal(raw, &fix); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}

	for _, c := range fix.Cases {
		t.Run(c.Name, func(t *testing.T) {
			s := DefaultSplitter()
			if c.Splitter != nil {
				s = *c.Splitter
			}
			got := SplitSentences(c.Input, s)
			if len(got) != len(c.Expected) {
				t.Fatalf("len mismatch: got %d, want %d (%v)", len(got), len(c.Expected), got)
			}
			for i, want := range c.Expected {
				if got[i].Text != want.Text {
					t.Errorf("case %q sent %d text: got %q, want %q", c.Name, i, got[i].Text, want.Text)
				}
				if got[i].CharOffset != want.CharOffset {
					t.Errorf("case %q sent %d offset: got %d, want %d", c.Name, i, got[i].CharOffset, want.CharOffset)
				}
			}
		})
	}
}

File: app/test/service/tts/sentence_splitter_parity_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:omnigram/service/tts/sentence_splitter.dart';

/// Parity test against `server/service/tts/testdata/splitter_parity.json`.
///
/// Both the Go splitter and this Dart port load the same fixture; if either
/// side changes the algorithm both tests must update in lockstep. That's
/// what keeps pre-gen alignment indices line up with client-side sentences.
void main() {
  test('splitter output matches Go fixture', () async {
    final fixtureFile = File('../server/service/tts/testdata/splitter_parity.json');
    expect(
      fixtureFile.existsSync(),
      isTrue,
      reason: 'fixture not found at ${fixtureFile.absolute.path}',
    );

    final raw = jsonDecode(await fixtureFile.readAsString()) as Map<String, dynamic>;
    final cases = raw['cases'] as List<dynamic>;
    expect(cases, isNotEmpty);

    for (final c in cases) {
      final m = c as Map<String, dynamic>;
      final name = m['name'] as String;
      final input = m['input'] as String;
      final splitterJson = m['splitter'] as Map<String, dynamic>?;
      final splitter = Splitter(
        lang: (splitterJson?['lang'] as String?) ?? '',
        minChars: (splitterJson?['minChars'] as int?) ?? 4,
        maxChars: (splitterJson?['maxChars'] as int?) ?? 200,
      );
      final expected = (m['expected'] as List<dynamic>)
          .map((e) => Sentence(
                text: (e as Map<String, dynamic>)['text'] as String,
                charOffset: e['charOffset'] as int,
              ))
          .toList();

      final got = splitSentences(input, splitter);
      expect(
        got.length,
        expected.length,
        reason: 'case "$name" length mismatch — got: $got',
      );
      for (var i = 0; i < expected.length; i++) {
        expect(got[i].text, expected[i].text, reason: 'case "$name" sent $i text');
        expect(got[i].charOffset, expected[i].charOffset,
            reason: 'case "$name" sent $i charOffset');
      }
    }
  });
}
